### PR TITLE
Adding step size for finite differences method, adjusting kwargs in QNodes

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
In certain cases, it would be nice if users would be able to specify the step size that is used for finite differences.

**Description of the Change:**
Adds a step size keyword argument to the `qnode` decorator, `QNode` and `JacobianQNode`. Further adds default values for **analytic** (`1e-7`) and **non-analytic** (`0.3`) device executions by choosing values that correspond to low absolute mean errors based on numerical results.

**Benefits:**
Users can pick their step sizes and can set it as well.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: pennylane/_device.py [hunk 1]
```diff
@@ -52,8 +52,9 @@ class Device(abc.ABC):
     _circuits = {}  #: dict[str->Circuit]: circuit templates associated with this API class
     _asarray = staticmethod(np.asarray)
 
-    def __init__(self, wires=1, shots=1000):
+    def __init__(self, wires=1, shots=1000, analytic=True):
```
Comment (josh146) [inline pennylane/_device.py]: This was something that came up previously, and we made the decision to not include `analytic` in the base Device class (see https://github.com/XanaduAI/pennylane/pull/317/files#r325266577)

File: pennylane/qnodes/decorator.py [hunk 2]
```diff
@@ -28,7 +28,7 @@
 ALLOWED_INTERFACES = ("autograd", "numpy", "torch", "tf")
 
 
-def QNode(func, device, *, interface="autograd", mutable=True, diff_method="best", properties=None):
+def QNode(func, device, *, interface="autograd", mutable=True, diff_method="best", h=None, properties=None):
```
Comment (josh146) [inline pennylane/qnodes/decorator.py]: As we add more QNodes, there will likely be a case where the QNodes take different keyword arguments, and it will be unmaintainable to keep adding arguments to the constructor function. Instead, perhaps should do the following:

* The QNode signature accepts `**kwargs` or an `options` dictionary. This dictionary is passed directly to the QNode to parse.

Do people prefer `**kwargs` or `options`? I'm undecided

Comment (josh146) [reply pennylane/qnodes/decorator.py]: It also avoids us having to update the constructor _and_ the decorator each time 😆

Comment (josh146) [reply pennylane/qnodes/decorator.py]: It also avoids Sphinx displaying the default value as `None`, which might be confusing to users

File: pennylane/qnodes/jacobian.py [hunk 3]
```diff
@@ -23,25 +23,54 @@
...
 
-    def __init__(self, func, device, mutable=True, properties=None):
+    def __init__(self, func, device, mutable=True, h=None, properties=None):
         super().__init__(func, device, mutable=mutable, properties=properties)
 
         self.par_to_grad_method = None
         """dict[int, str]: map from flattened quantum function positional parameter index
         to the gradient method to be used with that parameter"""
 
+        self._h = h or default_step_size(device.analytic)
```
Comment (josh146) [inline pennylane/qnodes/jacobian.py]: Nitpick: I find it more readable to have this logic be all in the same place:
```python
default_step_size = DEFAULT_STEP_SIZE_ANALYTIC if analytic else DEFAULT_STEP_SIZE
self._h = h or default_step_size 
```
I'm not sure that this logic is complex enough to warrant the overhead of a separate function.

(At some point it may be the case that the logic for finding the default _is_ more complicated, in which case we should move it to its own function, but I feel it is premature to do it now).

File: pennylane/qnodes/jacobian.py [hunk 4]
```diff
@@ -295,7 +328,7 @@ def _pd_finite_diff(self, idx, args, kwargs, **options):
             array[float]: partial derivative of the node
         """
         y0 = options.get("y0", None)
-        h = options.get("h", 1e-7)
+        h = options.get("h", self.h)
```
Comment (josh146) [inline pennylane/qnodes/jacobian.py]: Nice :)

File: pennylane/qnodes/jacobian.py [hunk 5]
```diff
@@ -174,6 +203,10 @@ def jacobian(self, args, kwargs=None, *, wrt=None, method="best", options=None):
 
         options = options or {}
 
+        # Add the step size into the options, if it was not there already
+        if "h" not in options.keys() and not self.device.analytic:
+            options = {"h": self.h, **options}
```
Comment (josh146) [inline pennylane/qnodes/jacobian.py]: Perfect 🎖

File: tests/qnodes/test_qnode_decorator.py [hunk 6]
```diff
@@ -90,7 +94,30 @@ def circuit(a):
...
+        return qml.expval(qml.PauliZ(wires=0))
+
+    assert not isinstance(circuit, CVQNode)
+    assert not isinstance(circuit, QubitQNode)
+    assert isinstance(circuit, JacobianQNode)
+    assert hasattr(circuit, "jacobian")
+    assert circuit.h == step_size
 
+    circuit.h = new_step_size
+    assert circuit.h == new_step_size
```
Comment (josh146) [inline tests/qnodes/test_qnode_decorator.py]: 💯

File: pennylane/qnodes/jacobian.py [hunk 7]
```diff
@@ -23,25 +23,54 @@
 
 from .base import BaseQNode, QuantumFunctionError
 
+DEFAULT_STEP_SIZE = 0.3
```
Comment (josh146) [inline pennylane/qnodes/jacobian.py]: Out of curiosity: what was the heuristic that led to a value of 0.3?

File: pennylane/qnodes/decorator.py [hunk 8]
```diff
@@ -79,6 +79,9 @@ def QNode(func, device, *, interface="autograd", mutable=True, diff_method="best
 
             * ``None``: a non-differentiable QNode is returned.
 
+        h (float): step size for parameter shift or the finite
+            difference method
```
Comment (josh146) [inline pennylane/qnodes/decorator.py]:
```suggestion
        h (float): step size for the finite-difference method
```
Note that this currently doesn't change the step-size for the parameter shift method --- that will likely require more work in a separate PR, as changing the divisor in the parameter shift method also involves making a corresponding change to the shift itself (so `QubitDevice._pd_analytic` would need to be modified)

File: pennylane/qnodes/decorator.py [hunk 9]
```diff
@@ -178,6 +181,8 @@ def qnode(device, *, interface="autograd", mutable=True, diff_method="best", pro
             * ``"finite-diff"``: Uses numerical finite-differences.
 
             * ``None``: a non-differentiable QNode is returned.
+        h (float): step size for parameter shift or the finite
+            difference method
```
Comment (josh146) [inline pennylane/qnodes/decorator.py]: Same comment here re: parameter shift

File: pennylane/_device.py (hunk 1 above)
Comment (josh146) [reply pennylane/_device.py]: (was there something here that necessitated adding this attribute?)

File: pennylane/qnodes/decorator.py (hunk 8 above)
Comment (josh146) [reply pennylane/qnodes/decorator.py]: Another thought: the default value of `h` should be specified here as well.

File: pennylane/_device.py (hunk 1 above)
Comment (antalszava) [reply pennylane/_device.py]: Oh, that makes sense! This was mainly added such that the `device.analytic` attribute can be checked later (and that separate cases can be created for the default step size). From what I experienced, when the step size was increased to a higher value in the `analytic=True` case (such as `0.3`), tests for the gradient computation had bigger deviations and were failing. Would there be perhaps a better way around this?

Comment (josh146) [reply pennylane/_device.py]: Oh, is this due to a testing issue? In tests, you are testing using a mocked out Device (where analytic doesn't exist) compared to a mocked out Default qubit?

File: pennylane/qnodes/decorator.py (hunk 2 above)
Comment (antalszava) [reply pennylane/qnodes/decorator.py]: Sounds good to me! Should that then mean that `properties` would also be passed as `kwargs`/`options` in this case and be passed on to the corresponding `QNode` that is being created?

File: pennylane/qnodes/decorator.py (hunk 8 above)
Comment (antalszava) [reply pennylane/qnodes/decorator.py]: Oh, right, thanks for getting that! :)

Would you suggest to include the default value of `h` here in the docstring or just in general in `decorator.py`?

File: pennylane/qnodes/decorator.py (hunk 2 above)
Comment (josh146) [reply pennylane/qnodes/decorator.py]: Oh good catch! We already had a options dictionary, it was just called properties 😆 In this case, we should use the existing `properties`, although I'm not sure I like the name.

What do you like in terms of naming? Since I don't think `properties` is being used, now is the perfect time to rename it if we want.

File: pennylane/qnodes/jacobian.py (hunk 3 above)
Comment (antalszava) [reply pennylane/qnodes/jacobian.py]: Sounds good to me :)

Comment (antalszava) [reply pennylane/qnodes/jacobian.py]: (The `device.analytic` attribute would be used here for the branching logic.)

File: pennylane/qnodes/decorator.py (hunk 8 above)
Comment (josh146) [reply pennylane/qnodes/decorator.py]: The default value should go in the docstring, whereever `h` is defined :)

File: pennylane/qnodes/decorator.py (hunk 2 above)
Comment (antalszava) [reply pennylane/qnodes/decorator.py]: I don't mind either :) I'll go with `kwargs` for now, but I'm up for any names that you think could make sense

File: pennylane/_device.py (hunk 1 above)
Comment (antalszava) [reply pennylane/_device.py]: Yes, several tests fail (mostly from the CV case, but there are a couple of qubit based ones using the `qubit_device_2_wires` fixture from `conftest.py`, ~90 test cases affected).

Another alternative I gave a try was adding the possibility to set the step size only in the `analytic=False` case (this was using the condition of `and not self.device.analytic` on line 207). This way, as you were suggesting:
* only the mocked out `Device` objects would require an `analytic` attribute (which could be made up for in the tests, ~10 test cases are affected overall)
* we wouldn't have to add the `analytic` attribute in the `Device` class
* however, the step size would not be settable in the `analytic=True` case

From my discussion with @trbromley, it seemed that the ability to set the step size  in the `analytic=True` case would not be too crucial, but rather something nice to have.

(Sidenote: the condition of `and not self.device.analytic` on line 207 was mistakenly left in for now, since we have a branching logic for the default values based on `device.analytic`, this should be removed using this logic.)

File: pennylane/qnodes/decorator.py (hunk 2 above)
Comment (trbromley) [reply pennylane/qnodes/decorator.py]: I like `**kwargs`!

File: pennylane/qnodes/jacobian.py (hunk 7 above)
Comment (trbromley) [reply pennylane/qnodes/jacobian.py]: This was just based on some initial numerics - I think what's key is to get the right order of magnitude. We can then update the number if we have more confidence on what it is.

Would it be possible for the default to depend on the number of shots?

**Comment by antalszava** [comment]
(Locally there seems to be no problem with blacking :thinking: )

File: pennylane/qnodes/base.py [hunk 10]
```diff
@@ -158,11 +158,11 @@ class BaseQNode:
             and returning a tuple of measured :class:`~.operation.Observable` instances.
         device (~pennylane._device.Device): computational device to execute the function on
         mutable (bool): whether the circuit is mutable, see above
-        properties (dict[str, Any] or None): additional keyword properties for adjusting the QNode behavior
+        kwargs (dict[str, Any] or None): additional keyword arguments for adjusting the QNode behavior
```
Comment (josh146) [inline pennylane/qnodes/base.py]: Typically, I don't think we include `kwargs` in the `Args:` list. Instead, we tend to do the following:

* If there are `kwargs` that can affect the properties of this QNode, we list them (alongside defaults) in a new `Keyword Args:` section below `Args:`.

* If there are no `kwargs` that can affect this QNode, we typically don't list it at all!

File: pennylane/qnodes/decorator.py [hunk 11]
```diff
@@ -178,20 +180,18 @@ def qnode(device, *, interface="autograd", mutable=True, diff_method="best", pro
             * ``"finite-diff"``: Uses numerical finite-differences.
 
             * ``None``: a non-differentiable QNode is returned.
+    Kwargs:
```
Comment (josh146) [inline pennylane/qnodes/decorator.py]:
```suggestion

    Keyword Args:
```

File: pennylane/qnodes/decorator.py (hunk 9 above)
Comment (josh146) [reply pennylane/qnodes/decorator.py]:
```suggestion
        h (float): Step size for the finite difference method. Default is `1e-7` for analytic devices, or
        `0.3` for non-analytic devices (those that estimate expectation values with a finite number of shots).
```

File: pennylane/qnodes/decorator.py [hunk 12]
```diff
@@ -178,20 +180,18 @@ def qnode(device, *, interface="autograd", mutable=True, diff_method="best", pro
...
+        h (float): step size for parameter shift or the finite
+            difference method
 
-        properties (dict[str->Any]): additional keyword properties passed to the QNode
-    """
+        kwargs (dict[str->Any]): additional keyword arguments passed to the QNode
```
Comment (josh146) [inline pennylane/qnodes/decorator.py]: I would remove this one for now (all individual keyword arguments should instead be listed here when needed, as you've done with `h` above :) )

File: pennylane/qnodes/decorator.py [hunk 13]
```diff
@@ -178,20 +180,18 @@ def qnode(device, *, interface="autograd", mutable=True, diff_method="best", pro
...
     def qfunc_decorator(func):
         """The actual decorator"""
         return QNode(
-            func,
-            device,
-            interface=interface,
-            mutable=mutable,
-            diff_method=diff_method,
-            properties=properties,
+            func, device, interface=interface, mutable=mutable, diff_method=diff_method, **kwargs
```
Comment (josh146) [inline pennylane/qnodes/decorator.py]: I like that black made this one line 😆

File: pennylane/qnodes/jacobian.py (hunk 7 above)
Comment (josh146) [reply pennylane/qnodes/jacobian.py]: Yep, we should definitely have the default depend on the number of shots once we know the relationship!

File: pennylane/qnodes/jacobian.py [hunk 14]
```diff
@@ -23,25 +23,47 @@
...
 
         self.par_to_grad_method = None
         """dict[int, str]: map from flattened quantum function positional parameter index
         to the gradient method to be used with that parameter"""
 
+        try:
+            analytic = self.device.analytic
+        except AttributeError:
+            # The analytic attribute is not defined for hardware devices
+            analytic = False
```
Comment (josh146) [inline pennylane/qnodes/jacobian.py]: Nice!

Comment (josh146) [reply pennylane/qnodes/jacobian.py]: Nitpick: this could also be written
```python
analytic = getattr(self.device, "analytic", False)
```
This is similar to the `get` method of dictionaries --- if the attribute does not exist, the default value (False) is returned.

Marked as a nitpick because try-except is also very Pythonic, so up to you whether you want to change this!

File: pennylane/qnodes/jacobian.py [hunk 15]
```diff
@@ -23,25 +23,47 @@
...
+
+        default_step_size = DEFAULT_STEP_SIZE_ANALYTIC if analytic else DEFAULT_STEP_SIZE
+        self._h = kwargs.get("h", default_step_size)
+        """float: step size for the finite difference method"""
```
Comment (josh146) [inline pennylane/qnodes/jacobian.py]: nice, we should definitely do more attribute docstrings where it is needed!

File: pennylane/qnodes/passthru.py [hunk 16]
```diff
@@ -69,14 +69,14 @@ class PassthruQNode(BaseQNode):
             A Python function containing :class:`~.operation.Operation` constructor calls,
             and returning a tuple of measured :class:`~.operation.Observable` instances.
         device (~pennylane._device.Device): computational device to execute the function on
-        properties (dict[str, Any] or None): additional keyword properties for adjusting the QNode behavior
+        kwargs (dict[str, Any] or None): additional keyword arguments for adjusting the QNode behavior
```
Comment (josh146) [inline pennylane/qnodes/passthru.py]: I would do the same thing you had in the decorator; remove this line, and have in it's place
```python
Keyword Args:
    use_native_type (bool): I don't know what this keyword argument does, but we should document it
```

File: pennylane/qnodes/base.py (hunk 10 above)
Comment (josh146) [reply pennylane/qnodes/base.py]: In this case, there are the `vis_check` and the `par_check` keyword arguments, so we could have:
```python
Keyword Args:
    vis_check (float): ...
    par_check (float): ...
```
Note that with the keyword args section, we _do_ provide default values in the description, since sphinx will not be able to automatically parse the defaults as it does with Args:.

**Comment by josh146** [comment]
According to the logs on the Black check, it looks like the issue is `qnodes/base.py`.

This could mean the version of black you have locally is out of date, you could try doing `pip install black --upgrade` and see if that fixes it

**Comment by josh146** [comment]
@antalszava just double checking the status of this PR

File: pennylane/qnodes/decorator.py (hunk 12 above)
Comment (antalszava) [reply pennylane/qnodes/decorator.py]: Sounds good, I'm removing this!

One thing here: would we still want to include the individual keyword arguments here (such as `h`)? As in the `BaseQNode` we would be omitting them, not sure if it would make sense to have them in the decorator.

File: pennylane/qnodes/base.py (hunk 10 above)
Comment (antalszava) [reply pennylane/qnodes/base.py]: Thanks so much! I've added these too (set the type to `bool` based on their usage)

File: pennylane/qnodes/jacobian.py (hunk 14 above)
Comment (antalszava) [reply pennylane/qnodes/jacobian.py]: I think this looks better, thanks so much for it! :blush: Also included a better-formatted attribute docstring

File: pennylane/qnodes/passthru.py (hunk 16 above)
Comment (antalszava) [reply pennylane/qnodes/passthru.py]: (_It's good that I double-checked the content here, almost went for just including it haha :smile:_)

**_tl;dr:_**
The only device using this right now is `default_tensor_tf` that uses this attribute for returning results without calling `Tensor.numpy()`.

The place where this is documented is in the `Device`  class as `default.tensor.tf`subclasses this along the line of inheritance.

I've used docstring from there to add it for:
* the `BaseQNode` with a default of `False`
* the `PassthruQNode` with a default of `True`

**_More details:_**

So this one seems to be a tough one, it seems to be used when passing the `return_native_type` argument in `base.py` to `device.execute`.

Ran two commands locally:
1. First looking for the occurences of `use_native_type`:
`grep -rnw . -e 'use_native_type' --include \*.py`
```python
./pennylane/pennylane/qnodes/base.py:775:        temp = self.kwargs.get("use_native_type", False)
./pennylane/pennylane/qnodes/passthru.py:77:        kwargs.setdefault("use_native_type", True)
```
2. Then looking for `return_native_type`:
`grep -rnw . -e 'return_native_type' --include \*.py`
```python
./pennylane/pennylane/_device.py:165:            return_native_type (bool): If True, return the result in whatever type the device uses
./pennylane/pennylane/qnodes/base.py:778:            ret = self.device.execute(self.circuit, return_native_type=temp)
./pennylane/pennylane/qnodes/base.py:784:                return_native_type=temp,
./pennylane/pennylane/beta/plugins/default_tensor_tf.py:356:        if kwargs.get("return_native_type", False):
```

These are results with having most plugins locally as well from the directory where the command was run.

**Comment by antalszava** [comment]
Just updated this based on your comments, thanks so much @josh146! :blush: I was uncertain about one thing, namely how we would like to include keyword arguments in `decorator.py`.

File: pennylane/qnodes/decorator.py (hunk 12 above)
Comment (josh146) [reply pennylane/qnodes/decorator.py]: Definitely good to have them in the decorator! The decorator is the docstring that is user facing, it is very unlikely the user will be looking at the BaseQNode docstring :)

File: pennylane/qnodes/passthru.py (hunk 16 above)
Comment (josh146) [reply pennylane/qnodes/passthru.py]: Thanks Antal! That is great, and helps reduce documentation debt

File: pennylane/_qubit_device.py [hunk 17]
```diff
@@ -140,6 +140,7 @@ def execute(self, circuit, **kwargs):
         Args:
             circuit (~.CircuitGraph): circuit to execute on the device
 
+
```
Comment (josh146) [inline pennylane/_qubit_device.py]:
```suggestion
```

File: pennylane/qnodes/base.py [hunk 18]
```diff
@@ -756,6 +759,10 @@ def evaluate(self, args, kwargs):
             args (tuple[Any]): positional arguments to the quantum function (differentiable)
             kwargs (dict[str, Any]): auxiliary arguments (not differentiable)
 
+        Keyword Args:
+            use_native_type (bool): If True, return the result in whatever type the device uses
+                internally, otherwise convert it into array[float]. Default: False.
```
Comment (josh146) [inline pennylane/qnodes/base.py]: Very nice!

File: pennylane/qnodes/decorator.py [hunk 19]
```diff
@@ -179,19 +180,16 @@ def qnode(device, *, interface="autograd", mutable=True, diff_method="best", pro
 
             * ``None``: a non-differentiable QNode is returned.
 
-        properties (dict[str->Any]): additional keyword properties passed to the QNode
-    """
+    Keyword Args:
+        h (float): Step size for the finite difference method. Default is `1e-7` for analytic devices, or
+            `0.3` for non-analytic devices (those that estimate expectation values with a finite number of shots).
```
Comment (josh146) [inline pennylane/qnodes/decorator.py]:
```suggestion
        h (float): Step size for the finite difference method. Default is ``1e-7`` for analytic devices, or
            ``0.3`` for non-analytic devices (those that estimate expectation values with a finite number of shots).
```
